### PR TITLE
Repeatedly attempt to start sled-agents in TQ bg task

### LINKED ISSUE #11032: Allow retries of the sending of `StartSledAgentRequest` from TQ background task
During the implementation of trust quorum, we discussed how to trigger starting sled-agents once the new sled was a member of the trust quorum. We thought that requests could not safely be resent to sled-agent and so we made sure that only a single nexus, in a single background task, could send this request exactly once. That could only happen by the nexus that transitioned the state of the trust quorum to `committed` - meaning all nodes have acknowledged the commit. And this send would only happen if the sled was new between last committed configuration and the latest configuration. 

We ended up recently with a situation where all but one of the sleds acknowledged a commit of trust quorum in epoch 6. One sled was permanently offline. So the situation was that we had a newly added sled in epoch 6, but we were in the `committing` state. This is safe, and all acked sleds are already operating in epoch 6 as far as trust quorum is concerned. This state solely exists to tell nexus to retry indefinitely to get an ack from the offline sled. It can be skipped with a new reconfiguration.

However,  we only start the sled-agent on the new sled once all nodes have acked. In this case that offline sled would never ack, and so we removed if from the trust quorum with a reconfiguration. Because there were enough committed nodes, trust quorum transitioned the state from `committing` to `committed-partially` and then reconfigured without the bad sled in epoch 7. All sleds acked in epoch 7. Unfortunately since the new sled was a member of *both* epoch 6 and 7, it was not detected as new by the background task and so the request to start sled-agent was not sent at commit time of epoch 7 by the background task that committed. In essence this diff assumed that the sled being added in 6 meant that sled-agent had properly started. But we couldn't retry to start the sled-agent anyway even if we didn't see the sled in the database, because we didn't think it was safe to do so.

In summary: we ended up in a state where a new sled was part of the trust quorum, but there was no sled-agent running on that sled, and the system would no longer try to start that sled-agent.

 It should be noted, that since we only try once, this could also happen if that request is lost in the network, or if the sled crashed transiently. We'd still be in this situation. One way out of this is to remove the new sled from trust quorum and then add it back again. Unfortunately, we can't do that because we key the removal operation on a `SledId` which doesn't exist. This `SledId` doesn't exist because its sled-agent was never started and it never upserted itself to Nexus such that an entry exists in CRDB. #11030 was opened to allow that to work.

However, @jgallagher also realized that starting a sled agent with a `StartSledAgentRequest` is actually idempotent. Therefore what we actually want to do is to change the trust quorum background task such that for the latest configuration that has committed enough sleds to commit partially, we repeatedly send a `StartSledAgentRequest` on a timer to any member of the trust quorum that does not yet have a corresponding row in the `sled` table. This will make the system recover on it's own in cases like the one we have here.

Note however, that if sled-agent fails to start correctly for any reason, we still need the escape hatch in #11030 to remove the sled from trust quorum and start again.

Also note, that we allocate a subnet for the new sled in the background task. That should also be made idempotent by checking if there is already a `BaseboardId` with a subnet.

With both this and #11030 fixed, we should not ever be blocked by failing to start sled-agent again.

### PR DESCRIPTION
**Pull request by andrewjstone**
This changes the strategy for starting sled-agents when adding a new sled via a trust quorum reconfiguration. It relies on the idempotence of starting a sled agent, and diffs the set of committed members in the latest TQ config with the non-expunged sleds in the sled table to find members that may need sled-agents started.

Fixes #11032

### CODE REVIEW

**Comment by andrewjstone** [comment]
I'm in the process of testing this.

File: nexus/db-queries/src/db/datastore/trust_quorum.rs [hunk 1]
```diff
@@ -1449,18 +1471,70 @@ impl DataStore {
...
+    async fn tq_get_committed_members_without_active_sled_conn(
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        rack_id: RackUuid,
+    ) -> Result<BTreeSet<BaseboardId>, TransactionError<Error>> {
+        use nexus_db_schema::schema::hw_baseboard_id::dsl as hw_dsl;
+        use nexus_db_schema::schema::sled::dsl as sled_dsl;
+        use nexus_db_schema::schema::trust_quorum_member::dsl;
+
+        let Some(latest) =
+            Self::tq_get_latest_config_conn(conn, rack_id).await?
```
Comment (jgallagher) [inline nexus/db-queries/src/db/datastore/trust_quorum.rs:1484]: Doing this as a separate query feels a little fishy to me: if we're not in a transaction a new config could be inserted between this query and the next. I'm not sure that matters here, but I'm also not sure it's any harder to make this part of the main query. If we do this:

```diff
diff --git a/nexus/db-queries/src/db/datastore/trust_quorum.rs b/nexus/db-queries/src/db/datastore/trust_quorum.rs
index 4afa8812a..abb0c9ce2 100644
--- a/nexus/db-queries/src/db/datastore/trust_quorum.rs
+++ b/nexus/db-queries/src/db/datastore/trust_quorum.rs
@@ -1476,19 +1476,20 @@ impl DataStore {
         conn: &async_bb8_diesel::Connection<DbConnection>,
         rack_id: RackUuid,
     ) -> Result<BTreeSet<BaseboardId>, TransactionError<Error>> {
+        use nexus_db_schema::schema::trust_quorum_configuration::dsl as cfg_dsl;
         use nexus_db_schema::schema::hw_baseboard_id::dsl as hw_dsl;
         use nexus_db_schema::schema::sled::dsl as sled_dsl;
         use nexus_db_schema::schema::trust_quorum_member::dsl;

-        let Some(latest) =
-            Self::tq_get_latest_config_conn(conn, rack_id).await?
-        else {
-            return Ok(BTreeSet::new());
-        };
+        let latest_epoch = cfg_dsl::trust_quorum_configuration
+            .select(cfg_dsl::epoch)
+            .filter(cfg_dsl::rack_id.eq(DbTypedUuid::<RackKind>::from(rack_id)))
+            .limit(1)
+            .order_by(cfg_dsl::epoch.desc());

         let rows: Vec<(String, String)> = dsl::trust_quorum_member
             .filter(dsl::rack_id.eq(DbTypedUuid::<RackKind>::from(rack_id)))
-            .filter(dsl::epoch.eq(latest.epoch))
+            .filter(dsl::epoch.eq_any(latest_epoch))
             .filter(dsl::state.eq(DbTrustQuorumMemberState::Committed))
             .inner_join(
                 hw_dsl::hw_baseboard_id.on(hw_dsl::id.eq(dsl::hw_baseboard_id)),
diff --git a/nexus/db-schema/src/schema.rs b/nexus/db-schema/src/schema.rs
index 336908bdf..c701f0308 100644
--- a/nexus/db-schema/src/schema.rs
+++ b/nexus/db-schema/src/schema.rs
@@ -3508,7 +3508,11 @@ table! {
     }
 }

-allow_tables_to_appear_in_same_query!(trust_quorum_member, hw_baseboard_id);
+allow_tables_to_appear_in_same_query!(
+    trust_quorum_member,
+    hw_baseboard_id,
+    trust_quorum_configuration
+);
 joinable!(trust_quorum_member -> hw_baseboard_id(hw_baseboard_id));

 // Declared as separate pairs rather than one three-table invocation, which
```

then we get this added as another `WHERE` clause:

```sql
("trust_quorum_member"."epoch" = ANY(
  SELECT "trust_quorum_configuration"."epoch" FROM "trust_quorum_configuration"
    WHERE ("trust_quorum_configuration"."rack_id" = $2)
    ORDER BY "trust_quorum_configuration"."epoch" DESC LIMIT $3
))
```

It does mean now we're duplicating the "get latest epoch" query with `tq_get_latest_config_conn`, but presumably that could be factored out to a method that returns the raw query they both use?

File: nexus/db-queries/src/db/datastore/trust_quorum.rs [hunk 2]
```diff
@@ -2348,4 +2449,91 @@ mod tests {
...
+        expected.insert(hw_ids[1].clone().into());
+        assert_eq!(missing_members().await, expected);
+
+        // A sled for the right baseboard but on a different rack does not
+        // count as present.
+        insert_sled_for_baseboard(datastore, RackUuid::new_v4(), &hw_ids[9])
+            .await;
+        assert_eq!(missing_members().await, expected);
+
+        db.terminate().await;
```
Comment (jgallagher) [inline nexus/db-queries/src/db/datastore/trust_quorum.rs:2579]: Maybe worth adding a test that expects `missing_members()` to be empty when all sleds in the config also have `sled` rows?

File: nexus/src/app/background/tasks/trust_quorum.rs [hunk 3]
```diff
@@ -226,16 +225,28 @@ async fn drive_reconfiguration(
         bail!("failed to retrieve configuration: missing");
     };
 
+    // We may have committed all members, but some of them may not have started sled-agents yet.
+    //
+    // We continue to retry sending them a `StartSledAgentRequest` until it starts.
```
Comment (jgallagher) [inline nexus/src/app/background/tasks/trust_quorum.rs]: turbo-nit - wrap these to 80 cols?

File: nexus/src/app/background/tasks/trust_quorum.rs
Comment (jgallagher) [inline nexus/src/app/background/tasks/trust_quorum.rs]: Is "Skipping" here misleading now that we don't return, and will continue on to trying to start sled-agents?

File: nexus/src/app/background/tasks/trust_quorum.rs [hunk 4]
```diff
@@ -244,25 +255,50 @@ async fn drive_reconfiguration(
...
     }
 
-    // For each unacked node, need to send a `Commit` or `PrepareAndCommit'
-    //
-    // At this point we know that the configuration is active and we are not
-    // preparing. Therefore we must be committing.
-    commit(log, opctx, datastore, config).await
+    // At this point, We know we've committed at least some members in this or
+    // prior attempts and need to start their sled-agents. Get the results for
+    // this attempt, if they exist.
```
Comment (jgallagher) [inline nexus/src/app/background/tasks/trust_quorum.rs:269]: We queried for `added_sleds` _before_ this iteration called `prepare()` or `commit()` - is that right? This comment says "in this or prior attempts" which is what I'd expect, but does that mean we need to query for `added_sleds` later?

File: nexus/src/app/background/tasks/trust_quorum.rs [hunk 5]
```diff
@@ -244,25 +255,50 @@ async fn drive_reconfiguration(
...
+        epoch,
+        added_sleds,
+    )
+    .await
+    .with_context(|| {
+        format!(
+            "Failed to start sled agents for added sleds for \
+                    rack {rack_id}, epoch {epoch}"
+        )
+    })?;
```
Comment (jgallagher) [inline nexus/src/app/background/tasks/trust_quorum.rs:302]: Returning an error here discards `commit_results` - is that okay?

File: nexus/src/app/background/tasks/trust_quorum.rs
Comment (andrewjstone) [reply nexus/src/app/background/tasks/trust_quorum.rs]: Ha! I predicted this comment. I removed it and added it back at one point. I was planning to add more log lines after relating to starting sled-agents. But looking again, I think you are right. I'll remove it.

File: nexus/db-queries/src/db/datastore/trust_quorum.rs (hunk 1 above)
Comment (andrewjstone) [reply nexus/db-queries/src/db/datastore/trust_quorum.rs:1484]: > Doing this as a separate query feels a little fishy to me: if we're not in a transaction a new config could be inserted between this query and the next. I'm not sure that matters here

It doesn't matter here. The members are all coupled to the configuration by the `rack_id` and `epoch`, so you are guaranteed to read only the relevant values. The race of reading the latest epoch stays the same. You'd either read it the in the transaction or not, but you won't know if someone else added a new config after reading.

I actually find the existing code easier to read, although it is one more round trip. But if you feel strongly I can go ahead and change it.

File: nexus/src/app/background/tasks/trust_quorum.rs (hunk 4 above)
Comment (andrewjstone) [reply nexus/src/app/background/tasks/trust_quorum.rs:269]: Yes, you are right, and it's actually a very easy fix a this point. Great catch. I changed the layout of this method a bunch before I found something I liked. I should have moved this though.

File: nexus/src/app/background/tasks/trust_quorum.rs (hunk 3 above)
Comment (andrewjstone) [reply nexus/src/app/background/tasks/trust_quorum.rs]: It's a solid nit.

File: nexus/src/app/background/tasks/trust_quorum.rs (hunk 5 above)
Comment (andrewjstone) [reply nexus/src/app/background/tasks/trust_quorum.rs:302]: I think so. It's just a bg task result that disappears on the next call. This just mimics the current behavior. We can always get the status from `omdb nexus trust-quorum get-status` anyway.

File: nexus/src/app/background/tasks/trust_quorum.rs (hunk 3 above)
Comment (andrewjstone) [reply nexus/src/app/background/tasks/trust_quorum.rs]: Fixed in [8258b96](https://github.com/oxidecomputer/omicron/pull/11064/commits/8258b9625c61bbde352d01505c22e9826a38e8a1)

File: nexus/db-queries/src/db/datastore/trust_quorum.rs (hunk 2 above)
Comment (andrewjstone) [reply nexus/db-queries/src/db/datastore/trust_quorum.rs:2579]: Good call. Done in [8258b96](https://github.com/oxidecomputer/omicron/pull/11064/commits/8258b9625c61bbde352d01505c22e9826a38e8a1)

File: nexus/src/app/background/tasks/trust_quorum.rs
Comment (andrewjstone) [reply nexus/src/app/background/tasks/trust_quorum.rs]: Done in [8258b96](https://github.com/oxidecomputer/omicron/pull/11064/commits/8258b9625c61bbde352d01505c22e9826a38e8a1)

File: nexus/src/app/background/tasks/trust_quorum.rs (hunk 4 above)
Comment (andrewjstone) [reply nexus/src/app/background/tasks/trust_quorum.rs:269]: Fixed in [8258b96](https://github.com/oxidecomputer/omicron/pull/11064/commits/8258b9625c61bbde352d01505c22e9826a38e8a1)

File: nexus/db-queries/src/db/datastore/trust_quorum.rs (hunk 1 above)
Comment (jgallagher) [reply nexus/db-queries/src/db/datastore/trust_quorum.rs:1484]: I don't feel super strongly.

**Review by jgallagher — APPROVED** [review]
Changes LGTM; is/was your testing able to see what happens if we send the same start request to a single sled multiple times?

**Comment by andrewjstone** [comment]
> Changes LGTM; is/was your testing able to see what happens if we send the same start request to a single sled multiple times?

I'm still testing it.

**Comment by andrewjstone** [comment]
> > Changes LGTM; is/was your testing able to see what happens if we send the same start request to a single sled multiple times?
> 
> I'm still testing it.

I do indeed see many attempts. Currently holding g1 offline and trying to start g3 in voxel I'm getting this error repeatedly:

```
   status = {"PerRackStatus":{"errors":["Reconfiguration worker error for rack_id 60729cec-6312-43df-9c35-397b0ef24f2b, epoch 2: Failed to start sled agents for added sleds for rack 60729cec-6312-43df-9c35-397b0ef24f2b, epoch 2: No sled-agent clients could be retrieved for trust quorum configuration with rack_id 60729cec-6312-43df-9c35-397b0ef24f2b and epoch 2"],"statuses":[]}}
   ```
  
I'm trying to figure out why that's occurring, and whether it's related to my setup or to these code changes.

**Comment by andrewjstone** [comment]
> I'm trying to figure out why that's occurring, and whether it's related to my setup or to these code changes.

I'm pretty sure this is a bug I introduce. I think I know how to fix it. Just one more round of testing...

**Comment by andrewjstone** [comment]
> > I'm trying to figure out why that's occurring, and whether it's related to my setup or to these code changes.
> 
> I'm pretty sure this is a bug I introduce. I think I know how to fix it. Just one more round of testing...

Yup. I'm getting a client error, because I want to get a client for each existing sled and then pick one to use, but instead I'm trying to get a client just for the new sled, which won't work because it doesn't have a sled-agent to connect to yet.

**Comment by andrewjstone** [comment]
> > > I'm trying to figure out why that's occurring, and whether it's related to my setup or to these code changes.
> > 
> > 
> > I'm pretty sure this is a bug I introduce. I think I know how to fix it. Just one more round of testing...
> 
> Yup. I'm getting a client error, because I want to get a client for each existing sled and then pick one to use, but instead I'm trying to get a client just for the new sled, which won't work because it doesn't have a sled-agent to connect to yet.

I was able to start the new sled-agent with one node down in the configuration.

```
20:48:37.972Z INFO f5b40ce7-6b09-47a4-8f5a-2e49a1b94711 (ServerContext): Successfully started sled agent
    background_task = trust_quorum_manager
    baseboard_id = 913-0000019:2FAKE003
    bgtask = TrustQuorumManager
    file = nexus/src/app/background/tasks/trust_quorum.rs:640
    sled_agent_baseboard_id = 913-0000019:2FAKE001
    sled_id = 278d3afa-1522-4879-b986-f4716a61456d
20:48:37.973Z INFO f5b40ce7-6b09-47a4-8f5a-2e49a1b94711 (ServerContext): Trust Quorum status for rack_id: a0a1f7d9-c243-4ff0-ab9e-fa4ec7eb53a2, epoch: 2:
        Commit Status from client 913-0000019:2FAKE002:
          acked: 913-0000019:2FAKE001
          pending:
          errors:
        Commit Status from client 913-0000019:2FAKE001:
          acked: 913-0000019:2FAKE000
          pending:
          errors:
        Commit Status from client 913-0000019:2FAKE000:
          acked: 913-0000019:2FAKE002, 913-0000019:2FAKE003
          pending:
          errors:
        StartSledAgent Status from client 913-0000019:2FAKE001:
          started:
            baseboard_id: 913-0000019:2FAKE003, sled_id: 278d3afa-1522-4879-b986-f4716a61456d
    background_task = trust_quorum_manager
    bgtask = TrustQuorumManager
    file = nexus/src/app/background/tasks/trust_quorum.rs:122
    ```

**Comment by andrewjstone** [comment]
And on the newly added sled:

```
20:42:23.526Z INFO SledAgent (SledAgentMain): Loading Sled Agent: StartSledAgentRequest { generation: 0, schema_version: 1, body: StartSledAgentRequestBody { id: 278d3afa-1522-4879-b986-f4716a61456d (sled), rack_id: a0a1f7d9-c243-4ff0-ab9e-fa4ec7eb53a2 (rack), use_trust_quorum: true, is_lrtq_learner: false, subnet: Ipv6Subnet { net: Ipv6Net { addr: fd00:17:1:d21::, width: 64 } } } }
```
